server/client.go

```diff
@@ -1966,6 +2009,71 @@ func (c *client) canSubscribe(subject string) bool {
 	return allowed
 }
 
+func (c *client) canQueueSubscribe(subject, queue string) bool {
+	if c.perms == nil {
+		return true
+	}
+
+	allowed := true
+	if c.perms.sub.allow != nil {
+		r := c.perms.sub.allow.Match(subject)
+
+		// If there is an explicit sub permission it allows any
+		// queue subscription unless there are explicit rules
+		// involving queue subscriptions.
+		subAllowed := len(r.psubs) != 0
+		allQueuesAllowed := len(r.qsubs) == 0
+		allowed = subAllowed && allQueuesAllowed
+
+		// We first check for explicit permissions for queue subscriptions,
+		// we later do another check for all subscriptions instead.
+	Loop:
+		for _, qsub := range r.qsubs {
+			for _, qq := range qsub {
```

Comment:
Do not need this loop, all are the same, just need to check first [0] as long as len() >= 1.

Reply:
Wally, `r.qsubs` is `[][]*subscription` and it contains an array of queue groups. So r.qsubs[0] say contains an array of all members of the queue name "foo" while r.qsubs[1] contains an array of all members of queue "bar", etc.. If you need to do processing based on the queue group (not individual member properties) then you just need to look at `r.qsubs[i][0]`.

Reply:
Thanks, I see now. Ok removed the second `for range` here